src/libmongoc/tests/test-mongoc-secure-channel.c

```diff
@@ -0,0 +1,263 @@
+#include <mongoc/mongoc.h>
+
+#include "TestSuite.h"
+#include "test-libmongoc.h"
+
+#if defined(MONGOC_ENABLE_SSL_SECURE_CHANNEL)
+#include <mongoc/mongoc-stream-tls-secure-channel-private.h>
+#include <mongoc/mongoc-host-list-private.h> // _mongoc_host_list_from_string_with_err
+#include <mongoc/mongoc-client-private.h>    // mongoc_client_connect_tcp
+#include <mongoc/mongoc-log-private.h>       // _mongoc_log_get_handler
+#include <test-conveniences.h>
+#include <mlib/test.h>
+
+static bool
+connect_with_secure_channel_cred (mongoc_ssl_opt_t *ssl_opt, mongoc_shared_ptr cred_ptr, bson_error_t *error)
+{
+   mongoc_host_list_t host;
+   const int32_t connect_timout_ms = 10000;
+
+   *error = (bson_error_t) {0};
+
+   if (!_mongoc_host_list_from_string_with_err (&host, "localhost:27017", error)) {
+      return false;
+   }
+   mongoc_stream_t *tcp_stream = mongoc_client_connect_tcp (connect_timout_ms, &host, error);
+   if (!tcp_stream) {
+      return false;
+   }
+
+   mongoc_stream_t *tls_stream = mongoc_stream_tls_secure_channel_new_with_creds (tcp_stream, ssl_opt, cred_ptr);
+   if (!tls_stream) {
+      mongoc_stream_destroy (tcp_stream);
+      return false;
+   }
+
+   if (!mongoc_stream_tls_handshake_block (tls_stream, host.host, connect_timout_ms, error)) {
+      mongoc_stream_destroy (tls_stream);
+      return false;
+   }
+
+   mongoc_stream_destroy (tls_stream);
+   return true;
+}
+
+// Test a TLS stream can be create with shared Secure Channel credentials.
+static void
+test_secure_channel_shared_creds_stream (void *unused)
+{
+   BSON_UNUSED (unused);
+
+   bool ok;
+   bson_error_t error;
+   mongoc_ssl_opt_t ssl_opt = {.ca_file = CERT_TEST_DIR "/ca.pem", .pem_file = CERT_TEST_DIR "/client.pem"};
+   // Test with no sharing:
+   {
+      ok = connect_with_secure_channel_cred (&ssl_opt, MONGOC_SHARED_PTR_NULL, &error);
+      ASSERT_OR_PRINT (ok, error);
+   }
+
+   // Test with sharing:
+   {
+      mongoc_shared_ptr cred_ptr =
+         mongoc_shared_ptr_create (mongoc_secure_channel_cred_new (&ssl_opt), mongoc_secure_channel_cred_deleter);
+      ok = connect_with_secure_channel_cred (&ssl_opt, cred_ptr, &error);
+      ASSERT_OR_PRINT (ok, error);
+      // Use again.
+      ok = connect_with_secure_channel_cred (&ssl_opt, cred_ptr, &error);
+      ASSERT_OR_PRINT (ok, error);
+      mongoc_shared_ptr_reset_null (&cred_ptr);
+   }
```

Comment:
Given this test component has visibility of `mongoc_stream_tls_secure_channel_t::cred_ptr`, suggest adding postcondition assertions that a reused `cred_ptr` is actually reflected by the internal `cred_ptr` (points to the same object).